### ISSUE TITLE
Downloading of VS Code extension slows down workspace start

### ISSUE THREAD

**Issue by garagatyi** [eclipse-che/che#12434]
### Description
On each restart of a workspace, we clean up plugins and restart downloading of VS code extensions binaries from the VS Code marketplace. This takes time and slows down Che workspace start. 
We need to cache binaries somehow to make workspace start blazing fast.

**Comment by garagatyi** [comment]
@slemeur @l0rd @benoitf WDYT?

**Comment by benoitf** [comment]
I think it's like for broker, if plug-ins are the same we may have a mode to not start broker at all (and then reuse all what was done previously)

**Comment by garagatyi** [comment]
@benoitf we could have something like this. But we don't have it currently.

**Comment by skabashnyuk** [comment]
BTW what was the reason why plugins folder was wiped out? We don't use PVC for that?

**Comment by garagatyi** [comment]
It is wiped out to get rid of plugins that were removed from the workspace. And it is using PVC.
We need to come up with a solution that would remove a plugin when it is needed and still cache plugins that are not changed.

**Comment by l0rd** [comment]
@garagatyi wouldn't #12398 fix that?

**Comment by garagatyi** [comment]
I might if we decide not to do brokering at all then workspace plugins configuration is not changed and versions of plugins are not upgraded. But when it is changed or new plugin version is released next workspace start would clean up everything and all the binaries would be downloaded again. For all the plugins, not just updated ones.

**Comment by l0rd** [comment]
@garagatyi I don't understand your answer. Can you try be more explicit? Is your answer related to #12398?

**Comment by garagatyi** [comment]
@l0rd if we implement detection of lack of workspace plugins configuration changes since the last provisioning we can avoid downloading of binaries. This is probably what you are referring to.

Another case is when something has changed - a plugin is updated/added/removed. Then we would need to start downloading of binaries of all the plugins. Should we think how to address this or maybe it is not something we need to think about this stage when the first case is not yet implemented?

**Comment by garagatyi** [comment]
To clarify: refactoring of plugin broker flow would affect extensions configuration evaluation time but would not affect extensions downloading time.
We do re-download of all the binaries on every workspace restart. Even when plugins list hasn't changed. 
When user configure a lot of plugins it can take a lot of time. 
Possible solution is to:
- cache plugins configuration (could be done by #12398 )
- do not restart plugin brokering if plugins configuration is cached and plugins list hasn't changed, so binaries of all the configured plugins are supposed to be in place (not in a case of ephemeral workspace)
